Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp

```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+        message = jscValueCreate(jscContext, jsValueRef);
+    }
+
+    GRefPtr<JSCValue> message;
+    GRefPtr<JSCValue> replyValue;
+    GUniquePtr<GError> replyError;
+    WTF::Function<void(API::SerializedScriptValue*, const String&)> completionHandler;
+    int referenceCount;
+};
+
+G_DEFINE_BOXED_TYPE(WebKitUserScriptMessage, webkit_user_script_message, webkit_user_script_message_ref, webkit_user_script_message_unref)
+
+/**
+ * webkit_user_script_message_ref:
+ * @user_script_message*: A #WebKitUserScriptMessage
+ *
+ * Atomically increments the reference count of @user_script_message by one.
+ *
+ * Returns: the @user_script_message passed in.
+ *
+ * Since: 2.38
+ */
+WebKitUserScriptMessage* webkit_user_script_message_ref(WebKitUserScriptMessage* userScriptMessage)
+{
+    g_return_val_if_fail(userScriptMessage, nullptr);
+    g_atomic_int_inc(&userScriptMessage->referenceCount);
+    return userScriptMessage;
+}
+
+/**
+ * webkit_user_script_message_unref:
+ * @user_script_message*: A #WebKitUserScriptMessage
+ *
+ * Atomically decrements the reference count of @user_script_message by one.
+ *
+ * If the reference count drops to 0, all the memory allocated by the
+ * #WebKitUserScriptMessage is released. This function is MT-safe and may
+ * be called from any thread.
+ *
+ * Since: 2.38
+ */
+void webkit_user_script_message_unref(WebKitUserScriptMessage* userScriptMessage)
+{
+    g_return_if_fail(userScriptMessage);
+    if (g_atomic_int_dec_and_test(&userScriptMessage->referenceCount)) {
+        userScriptMessage->~WebKitUserScriptMessage();
+        fastFree(userScriptMessage);
+    }
+}
+
+WebKitUserScriptMessage* webKitUserScriptMessageCreate(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
+{
+    WebKitUserScriptMessage* userScriptMessage = static_cast<WebKitUserScriptMessage*>(fastMalloc(sizeof(WebKitUserScriptMessage)));
+    new (userScriptMessage) WebKitUserScriptMessage(serializedScriptValue, WTFMove(completionHandler));
+    return userScriptMessage;
+}
+
+/**
+ * webkit_user_script_message_get_value:
+ * @user_script_message: A #WebKitUserScriptMessage
+ *
+ * Return the message sent in by user.
+ *
+ * Returns: (transfer none): The #JSCValue sent by JavaScript code to the
+ * script message handler.
+ *
+ * Since: 2.38
+ */
+JSCValue* webkit_user_script_message_get_value(WebKitUserScriptMessage* userScriptMessage)
+{
+    return userScriptMessage->message.get();
+}
+
+/**
+ * webkit_user_script_message_reply:
+ * @user_script_message: A #WebKitUserScriptMessage
+ * @reply_value: Reply value of the provided script
+ *
+ * Provide the reply to the user provided script.
+ *
+ * Since: 2.38
+ */
+void webkit_user_script_message_reply(WebKitUserScriptMessage* message, JSCValue* replyValue)
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(JSC_IS_VALUE(replyValue));
+    message->replyValue = replyValue;
```

Comment:
If we go with the approach that I suggested above, we need to construct an `API::SerializedScriptValue` which represents the passed `replyValue`, call the `completionHandler`, and then set it to `nullptr` (to mark it as used/invalid). We can also add a check to make sure we get warnings if trying to use the `completionHandler` more than once:

```suggestion
    g_return_if_fail(JSC_IS_VALUE(replyValue));
    g_return_if_fail(message->completionHandler);

    auto serializedValue = API::SerializedScriptValue::create(...);
    message->completionHandler(serializedValue.get(), { });
    message->completionHandler = nullptr;
```